src/lib/components/availability/SetAvailabilityModal.svelte

```diff
@@ -0,0 +1,170 @@
+	import {
+		Modal,
+		Grid,
+		Row,
+		Column,
+		TextInput,
+		Checkbox,
+		TimePicker
+	} from 'carbon-components-svelte'
+	import { Edit } from 'carbon-icons-svelte'
+	import { collection, doc, setDoc, Timestamp } from 'firebase/firestore'
+
+	export let rehearsal: Rehearsal
+	export let user: { id: string; name: string }
+	export let availability: Availability | undefined = undefined
+
+	export let displayUserName: boolean = false
+
+	export let openSetAvailability = false
+	export let onClose: () => void
+
+	let setOrUpdate = availability ? 'Update' : 'Set'
+
+	let available = availability?.available ?? true
+	let remark: string | undefined = availability?.reason
+
+	let startTimeString: string = getTimeString(availability?.startTime ?? rehearsal.startTime)
+	let endTimeString: string = getTimeString(availability?.endTime ?? rehearsal.endTime)
+	let startTimeInvalid: boolean = false
+	let endTimeInvalid: boolean = false
+
+	/**
+	 * Submit or update the availability in the database.
+	 * @returns `true` if the mutation was successful, `false` otherwise.
+	 */
+	async function saveAvailability(): Promise<boolean> {
+		// Time input validation
+		const timeRegex = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/
+		startTimeInvalid = !timeRegex.test(startTimeString)
+		endTimeInvalid = !timeRegex.test(endTimeString)
+		if (startTimeInvalid || endTimeInvalid) return false
+
+		// If the participant is not available, they are unavailable for the entire rehearsal
+		if (!available) {
+			startTimeString = getTimeString(rehearsal.startTime)
+			endTimeString = getTimeString(rehearsal.endTime)
+		}
+
+		// Check if the availability is within the rehearsal time and if the start time is before the end time
+		const rehearsalStart = rehearsal.startTime.toDate()
+		const rehearsalEnd = rehearsal.endTime.toDate()
+
+		const start = startTimeString.split(':')
+		const end = endTimeString.split(':')
+
+		const availableStart = newTimeOnDay(rehearsalStart, parseInt(start[0]), parseInt(start[1]))
+		const availableEnd = newTimeOnDay(rehearsalEnd, parseInt(end[0]), parseInt(end[1]))
+
+		startTimeInvalid =
+			rehearsalStart > availableStart ||
+			rehearsalEnd < availableStart ||
+			availableStart >= availableEnd
+		endTimeInvalid =
+			rehearsalStart > availableEnd || rehearsalEnd < availableEnd || availableStart >= availableEnd
+
+		if (startTimeInvalid || endTimeInvalid) return false
+
+		// Save the availability to the database
+		const newAvailibilityDoc: NewAvailability = {
+			available,
+			rehearsal: doc(db, 'rehearsals/' + rehearsal.id),
+			reason: remark,
+			startTime: Timestamp.fromDate(availableStart),
+			endTime: Timestamp.fromDate(availableEnd)
+		}
+
+		const availabilityRef =
+			availability?.id !== undefined
+				? doc(db, 'users/' + user.id + '/availability/' + availability.id)
+				: doc(collection(db, 'users/' + user.id + '/availability'))
+
+		await setDoc(availabilityRef, newAvailibilityDoc)
+
+		invalidateAll()
+
+		return true
```

Comment:
isn't like most of this copied from somewhere else? If so just make it a pure function that you pass a client or smth, and reuse it. Will make the component easier to read too

Reply:
Well not anymore, because this is the only place now to save/update availability. I can still extract it to another file if you prefer

Reply:
Yes would be nice, we need to clean up a lot of these long code string imo